src/daemon/rpmostreed-transaction-livefs.c

```diff
@@ -0,0 +1,695 @@
+  g_autoptr(GPtrArray) added = g_ptr_array_new_with_free_func ((GDestroyNotify) g_object_unref);
+  if (!ostree_diff_dirs (0, from_tree, to_tree, modified, removed, added,
+                         cancellable, error))
+    return FALSE;
+
+  /* We'll filter these arrays below. */
+  diff->modified = g_ptr_array_new_with_free_func ((GDestroyNotify) g_object_unref);
+  diff->removed = g_ptr_array_new_with_free_func ((GDestroyNotify) g_object_unref);
+  diff->added = g_ptr_array_new_with_free_func ((GDestroyNotify) g_object_unref);
+
+  /* Analyze the differences */
+  for (guint i = 0; i < modified->len; i++)
+    {
+      OstreeDiffItem *diffitem = modified->pdata[i];
+      const char *path = gs_file_get_path_cached (diffitem->src);
+
+      if (diff_one_path (diff, path) == FILE_DIFF_RESULT_KEEP)
+        g_ptr_array_add (diff->modified, g_object_ref (diffitem->src));
+    }
+
+  for (guint i = 0; i < removed->len; i++)
+    {
+      GFile *gfpath = removed->pdata[i];
+      const char *path = gs_file_get_path_cached (gfpath);
+
+      if (diff_one_path (diff, path) == FILE_DIFF_RESULT_KEEP)
+        g_ptr_array_add (diff->removed, g_object_ref (gfpath));
+    }
+
+  for (guint i = 0; i < added->len; i++)
+    {
+      GFile *added_f = added->pdata[i];
+      const char *path = gs_file_get_path_cached (added_f);
+
+      if (diff_one_path (diff, path) == FILE_DIFF_RESULT_KEEP)
+        g_ptr_array_add (diff->added, g_object_ref (added_f));
+    }
+
+  /* And gather the RPM level changes */
+  if (!rpm_ostree_db_diff (repo, from_rev, to_rev,
+                           &diff->removed_pkgs, &diff->added_pkgs,
+                           &diff->modified_pkgs_old, &diff->modified_pkgs_new,
+                           cancellable, error))
+    return FALSE;
+  g_assert (diff->modified_pkgs_old->len == diff->modified_pkgs_new->len);
+
+  *out_diff = g_steal_pointer (&diff);
+  return TRUE;
+}
+
+static void
+print_commit_diff (CommitDiff *diff)
+{
+  /* Print out the results of the two diffs */
+  g_print ("Diff Analysis: %s => %s\n", diff->from, diff->to);
+  g_print ("Files:\n modified: %u\n removed: %u\n added: %u\n",
+           diff->modified->len, diff->removed->len, diff->added->len);
+  g_print ("Packages:\n modified: %u\n removed: %u\n added: %u\n",
+           diff->modified_pkgs_new->len, diff->removed_pkgs->len, diff->added_pkgs->len);
+
+  if (diff->flags & COMMIT_DIFF_FLAGS_ETC)
+    {
+      g_print ("* Configuration changed in /etc\n");
+    }
+  if (diff->flags & COMMIT_DIFF_FLAGS_ROOTFS)
+    {
+      g_print ("* Content outside of /usr and /etc is modified\n");
+    }
+  if (diff->flags & COMMIT_DIFF_FLAGS_BOOT)
+    {
+      g_print ("* Kernel/initramfs changed\n");
+    }
+  fflush (stdout);
+}
+
+/* We want to ensure the rollback deployment matches our booted checksum. If it
+ * doesn't, we'll push a new one, and GC the previous one(s).
+ */
+static OstreeDeployment *
+get_rollback_deployment (OstreeSysroot *self,
```

Comment:
We can simplify this further with the new `rpmostree_syscore_query_deployments()`, right?